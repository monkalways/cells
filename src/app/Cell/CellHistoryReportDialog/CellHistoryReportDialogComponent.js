import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { compose } from 'recompose';
import { withRouter } from 'react-router-dom';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  FormHelperText,
  Grid,
} from '@material-ui/core';
import DateFnsUtils from '@date-io/date-fns';
import { subDays } from 'date-fns';
import {
  MuiPickersUtilsProvider,
  TimePicker,
  DatePicker,
} from 'material-ui-pickers';

import Loading from '../../common/Loading';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  cellDetails: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  isLoadingReport: PropTypes.bool.isRequired,
  handleLoadReport: PropTypes.func.isRequired,
  handleCloseModal: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export class CellHistoryReportDialogComponent extends Component {
  state = {
    startTime: subDays(Date.now(), 1),
    endTime: Date.now(),
    error: null,
  };

  handleClose = () => {
    const { handleCloseModal } = this.props;
    handleCloseModal();
  };

  handleStartTimeChange = (startTime) => {
    const { endTime } = this.state;
    if (startTime >= endTime) {
      this.setState({
        error: 'Start Date/Time must be before End Date/Time.',
      });
      return;
    }
    this.setState({ startTime, error: null });
  };

  handleEndTimeChange = (endTime) => {
    const { startTime } = this.state;
    if (startTime >= endTime) {
      this.setState({
        error: 'End Date/Time must be after Start Date/Time.',
      });
      return;
    }
    this.setState({ endTime, error: null });
  };

  handleConfirm = () => {
    const {
      handleLoadReport,
      history,
      cellDetails,
      handleCloseModal,
    } = this.props;
    const { startTime, endTime } = this.state;
    handleLoadReport(cellDetails.name, startTime, endTime);
    handleCloseModal();
    history.push(`/cells/${cellDetails.name}/home/cell-history-report`);
  };

  render() {
    const { isModalOpen, isLoadingReport, classes } = this.props;
    const { startTime, endTime, error } = this.state;
    return (
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <Dialog
          open={isModalOpen}
          onClose={this.handleClose}
          disableBackdropClick={isLoadingReport}
          disableEscapeKeyDown={isLoadingReport}
        >
          <DialogTitle>
            Specify start and end times for the Cell History Report:
          </DialogTitle>
          {isLoadingReport ? (
            <DialogContent>
              <Loading size={50} />
            </DialogContent>
          ) : (
            <DialogContent>
              <Grid container justify="space-around">
                <Grid item xs={12}>
                  <FormControl className={classes.formControl}>
                    <DatePicker
                      autoOk
                      disableFuture
                      margin="normal"
                      label="Start Date"
                      format="MMM dd, yyyy"
                      variant="outlined"
                      value={startTime}
                      onChange={this.handleStartTimeChange}
                    />
                  </FormControl>
                  <FormControl className={classes.formControl}>
                    <TimePicker
                      autoOk
                      margin="normal"
                      label="Start Time"
                      variant="outlined"
                      value={startTime}
                      onChange={this.handleStartTimeChange}
                    />
                  </FormControl>
                </Grid>
                <Grid item xs={12}>
                  <FormControl className={classes.formControl}>
                    <DatePicker
                      autoOk
                      disableFuture
                      margin="normal"
                      format="MMM dd, yyyy"
                      label="End Date"
                      variant="outlined"
                      value={endTime}
                      onChange={this.handleEndTimeChange}
                    />
                  </FormControl>
                  <FormControl className={classes.formControl}>
                    <TimePicker
                      autoOk
                      margin="normal"
                      label="End Time"
                      variant="outlined"
                      value={endTime}
                      onChange={this.handleEndTimeChange}
                    />
                  </FormControl>
                </Grid>
                {error && (
                  <Grid item xs={12}>
                    <FormHelperText error className={classes.errorText}>
                      {error}
                    </FormHelperText>
                  </Grid>
                )}
              </Grid>
            </DialogContent>
          )}
          <DialogActions>
            <Button
              id="cancelButton"
              onClick={this.handleClose}
              color="primary"
              disabled={isLoadingReport}
            >
              Cancel
            </Button>
            <Button
              id="loadReportButton"
              onClick={this.handleConfirm}
              color="primary"
              variant="contained"
              autoFocus
              disabled={isLoadingReport}
            >
              Confirm
            </Button>
          </DialogActions>
        </Dialog>
      </MuiPickersUtilsProvider>
    );
  }
}

CellHistoryReportDialogComponent.propTypes = propTypes;

export default compose(
  withStyles((theme) => ({
    formControl: {
      margin: theme.spacing.unit,
    },
    errorText: {
      fontSize: theme.typography.fontSize * 1.2,
    },
  })),
  withRouter,
)(CellHistoryReportDialogComponent);
