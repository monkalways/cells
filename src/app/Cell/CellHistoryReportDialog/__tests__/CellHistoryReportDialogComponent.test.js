import React from 'react';
import { shallow } from 'enzyme';
import Dialog from '@material-ui/core/Dialog';
import Button from '@material-ui/core/Button';
import { CellHistoryReportDialogComponent } from '../CellHistoryReportDialogComponent';

describe('CellHistoryReportDialogComponent', () => {
  let name;
  let isModalOpen;
  let isLoadingReport;
  let handleLoadReport;
  let handleCloseModal;
  let push;

  beforeEach(() => {
    name = 'c1';
    handleLoadReport = jest.fn();
    handleCloseModal = jest.fn();
    isModalOpen = false;
    isLoadingReport = false;
    push = jest.fn();

    Date.now = jest.fn(() => new Date(Date.UTC(2019, 2, 12, 0)));
  });

  const setup = () => {
    const classes = {};
    return shallow(<CellHistoryReportDialogComponent
      classes={classes}
      cellDetails={{ name }}
      isModalOpen={isModalOpen}
      isLoadingReport={isLoadingReport}
      handleLoadReport={handleLoadReport}
      handleCloseModal={handleCloseModal}
      history={{ push }}
    />);
  };

  it('should render', () => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should set initial state', () => {
    const wrapper = setup();
    expect(wrapper.state().startTime).toEqual(new Date(Date.UTC(2019, 2, 11, 0)));
    expect(wrapper.state().endTime).toEqual(new Date(Date.UTC(2019, 2, 12, 0)));
    expect(wrapper.state().endTime - wrapper.state().startTime).toBe(8.64e7);
    expect(wrapper.state().error).toBeNull();
  });

  it('should handleClose', () => {
    const wrapper = setup();

    wrapper.find(Dialog).simulate('close');
    expect(handleCloseModal).toBeCalled();
  });

  it('should handleStartTimeChange with valid date change', () => {
    const wrapper = setup();
    const newStartTime = new Date(Date.UTC(2019, 2, 10, 0));

    wrapper.find({ label: 'Start Date' }).simulate('change', newStartTime);

    expect(wrapper.state().startTime).toBe(newStartTime);
    expect(wrapper.state().error).toBeNull();
  });

  it('should handleStartTimeChange with invalid date change', () => {
    const wrapper = setup();
    const newStartTime = new Date(Date.UTC(2019, 2, 13, 0));

    wrapper.find({ label: 'Start Date' }).simulate('change', newStartTime);

    expect(wrapper.state().startTime).toEqual(new Date(Date.UTC(2019, 2, 11, 0)));
    expect(wrapper.state().error).toBe('Start Date/Time must be before End Date/Time.');
  });

  it('should handleStartTimeChange with valid time change', () => {
    const wrapper = setup();
    const newStartTime = new Date(Date.UTC(2019, 2, 11, 1));

    wrapper.find({ label: 'Start Time' }).simulate('change', newStartTime);

    expect(wrapper.state().startTime).toBe(newStartTime);
    expect(wrapper.state().error).toBeNull();
  });

  it('should handleStartTimeChange with invalid time change', () => {
    const wrapper = setup();
    const newStartTime = new Date(Date.UTC(2019, 2, 12, 1));

    wrapper.find({ label: 'Start Time' }).simulate('change', newStartTime);

    expect(wrapper.state().startTime).toEqual(new Date(Date.UTC(2019, 2, 11, 0)));
    expect(wrapper.state().error).toBe('Start Date/Time must be before End Date/Time.');
  });

  it('should handleEndTimeChange with valid date change', () => {
    const wrapper = setup();
    const newEndTime = new Date(Date.UTC(2019, 2, 13, 0));

    wrapper.find({ label: 'End Date' }).simulate('change', newEndTime);

    expect(wrapper.state().endTime).toBe(newEndTime);
    expect(wrapper.state().error).toBeNull();
  });

  it('should handleEndTimeChange with invalid date change', () => {
    const wrapper = setup();
    const newEndTime = new Date(Date.UTC(2019, 2, 11, 0));

    wrapper.find({ label: 'End Date' }).simulate('change', newEndTime);

    expect(wrapper.state().endTime).toEqual(new Date(Date.UTC(2019, 2, 12, 0)));
    expect(wrapper.state().error).toBe('End Date/Time must be after Start Date/Time.');
  });

  it('should handleEndTimeChange with valid time change', () => {
    const wrapper = setup();
    const newEndTime = new Date(Date.UTC(2019, 2, 11, 1));

    wrapper.find({ label: 'End Time' }).simulate('change', newEndTime);

    expect(wrapper.state().endTime).toBe(newEndTime);
    expect(wrapper.state().error).toBeNull();
  });

  it('should handleEndTimeChange with invalid time change', () => {
    const wrapper = setup();
    const newEndTime = new Date(Date.UTC(2019, 2, 10, 1));

    wrapper.find({ label: 'End Time' }).simulate('change', newEndTime);

    expect(wrapper.state().endTime).toEqual(new Date(Date.UTC(2019, 2, 12, 0)));
    expect(wrapper.state().error).toBe('End Date/Time must be after Start Date/Time.');
  });

  it('should handleConfirm', () => {
    name = 'c2';
    const wrapper = setup();

    // last Button is the confirm button
    wrapper
      .find(Button)
      .last()
      .simulate('click');

    const expectedStartTime = new Date(Date.UTC(2019, 2, 11, 0));
    const expectedEndTime = new Date(Date.UTC(2019, 2, 12, 0));
    expect(handleLoadReport).toBeCalledWith(
      name,
      expectedStartTime,
      expectedEndTime,
    );
    expect(handleCloseModal).toBeCalled();
    expect(push).toBeCalledWith('/cells/c2/home/cell-history-report');
  });
});
