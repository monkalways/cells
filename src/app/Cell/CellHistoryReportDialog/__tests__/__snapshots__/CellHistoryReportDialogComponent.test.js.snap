// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellHistoryReportDialogComponent should render 1`] = `
<MuiPickersUtilsProvider
  utils={[Function]}
>
  <WithStyles(Dialog)
    disableBackdropClick={false}
    disableEscapeKeyDown={false}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(DialogTitle)>
      Specify start and end times for the Cell History Report:
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)>
      <WithStyles(Grid)
        container={true}
        justify="space-around"
      >
        <WithStyles(Grid)
          item={true}
          xs={12}
        >
          <WithStyles(FormControl)>
            <ForwardRef
              autoOk={true}
              disableFuture={true}
              format="MMM dd, yyyy"
              label="Start Date"
              margin="normal"
              onChange={[Function]}
              value={2019-03-11T00:00:00.000Z}
              variant="outlined"
            />
          </WithStyles(FormControl)>
          <WithStyles(FormControl)>
            <ForwardRef
              autoOk={true}
              label="Start Time"
              margin="normal"
              onChange={[Function]}
              value={2019-03-11T00:00:00.000Z}
              variant="outlined"
            />
          </WithStyles(FormControl)>
        </WithStyles(Grid)>
        <WithStyles(Grid)
          item={true}
          xs={12}
        >
          <WithStyles(FormControl)>
            <ForwardRef
              autoOk={true}
              disableFuture={true}
              format="MMM dd, yyyy"
              label="End Date"
              margin="normal"
              onChange={[Function]}
              value={2019-03-12T00:00:00.000Z}
              variant="outlined"
            />
          </WithStyles(FormControl)>
          <WithStyles(FormControl)>
            <ForwardRef
              autoOk={true}
              label="End Time"
              margin="normal"
              onChange={[Function]}
              value={2019-03-12T00:00:00.000Z}
              variant="outlined"
            />
          </WithStyles(FormControl)>
        </WithStyles(Grid)>
      </WithStyles(Grid)>
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        color="primary"
        disabled={false}
        id="cancelButton"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <WithStyles(Button)
        autoFocus={true}
        color="primary"
        disabled={false}
        id="loadReportButton"
        onClick={[Function]}
        variant="contained"
      >
        Confirm
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</MuiPickersUtilsProvider>
`;
